File: OJ/leet486/leet486.cpp
/*
leet486


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <random>
#include <functional>
#include <utility>
#include "../utils/LeetCpp.utils.hpp"

using namespace std;

class Solution {
public:
    bool PredictTheWinner(const vector<int>& A) {
        int n = A.size();
        vector<int> g = {0}; g.reserve(n+1);
        for (auto&& x: A) g.push_back(g.back() + x);

        vector<vector<int>> f(n, vector<int>(n));
        
        for (int i = 0; i < n; ++i) f[i][i] = A[i];
        for (int len = 1; len < n; len++) {
            for (int i = 0; i + len < n; i++) {
                auto sum1 = g[i+len+1] - g[i+1];
                auto sum2 = g[i+len] - g[i];
                f[i][i+len] = max(A[i] - f[i+1][i+len] + sum1, A[i+len] - f[i][i+len-1] + sum2);
            }
        }
        // cout << g << endl;
        // cout << f << endl;
        return f[0][n-1] * 2 >= g.back();
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << boolalpha;
    // cout << sol.PredictTheWinner({1,5,2,4,6}) << endl;
    // cout << sol.PredictTheWinner({100, 100, 100}) << endl;
    // cout << sol.PredictTheWinner({1, 5, 233, 7}) << endl;
    // cout << sol.PredictTheWinner({1, 5, 2}) << endl;
    cout << sol.PredictTheWinner({1, 3, 1}) << endl;
    cout << sol.PredictTheWinner({1, 2}) << endl;
    return 0;
}

File: OJ/leet486/leet486.py
# leet486

from typing import List

from copy import deepcopy 
from pprint import pprint
class Solution:
    def PredictTheWinner1(self, A: List[int]) -> bool:
        n = len(A)
        f = [[0]*n for _ in range(n)]
        for i in range(n):
            f[i][i] = A[i]
        for i in range(n-2,-1,-1):
            for j in range(i+1,n):
                f[i][j] = max(A[j]-f[i][j-1],A[i]-f[i+1][j])
        pprint(f)
        return f[0][-1] >= 0
    

    def PredictTheWinner(self, A: List[int]) -> bool:
        # f(i, j): A[i..j] 上先手比后手最多多拿的分数
        # f(i, j) = max( A[i] - f(i+1, j), A[j] - f(i, j-1) )
        n = len(A)
        memo = [[None]*n for _ in range(n)]
        for i in range(n): memo[i][i] = A[i]

        def f(i, j):
            if i > j: return 0
            if memo[i][j] is not None: return memo[i][j]

            memo[i][j] = max( A[i] - f(i+1, j), A[j] - f(i, j-1) )
            return memo[i][j]
        
        return f(0, n-1) >= 0



print(Solution().PredictTheWinner([0]))
print(Solution().PredictTheWinner([1, 5, 2]))
print(Solution().PredictTheWinner([1, 5, 233, 7]))
print(Solution().PredictTheWinner([0,0,7,6,5,6,1]))


